File: Go/math/Mystery_Number/Mystery_Number.go
/* Below checks checks if a number is a mystery number.
Mystery number - A number that can be represented by sum
of two numbers, such that the two numbers are reverse of
each other.*/

package main

import (
	"fmt"
	"strconv"
)

/* This function takes a number as a input and returns
a number with the digits reversed*/
func rev(number int) int {

	var reverse string
	for number > 0 {
		/* The number is broken into pieces by taking mod 10.
		Each digit is converted to string and concatenated.*/
		str := strconv.Itoa(number % 10)
		reverse += str
		number = number / 10
	}
	// Finally the reversed string is converted back to integer
	revs, err := strconv.Atoi(reverse)
	if(err == nil) {
		// If there were no errors we return the number
		return revs
	}
	return -1
}

// This function checks if the number is a mystery number
func mystery(number int) int {

	// We iterate till the middle of the number
	var distance int = number / 2
	for i:= 1; i <= distance; i++ {
		// Check the sum of number with its reverse
		if(i + rev(i) == number) {
			// If its is equal to the given number we return and print
			fmt.Print("\nThe two digits are: ", i, " ", rev(i))
			return 1
		}
	}
	// Else return -1
	return -1
}

func main() {

	// Taking number as input from the user
	fmt.Print("Enter a number to check if it's a mystery number: ")
	var number int
	fmt.Scan(&number)

	// Calling the mystery function
	var result int = mystery(number)

	// If the result is 1, the number is a mystery number
	if(result == 1) {
		fmt.Print("\nThe number ", number, " is a mystery number.\n")
	} else {
		// Else it is not a mystery number
		fmt.Print("\nThe number ", number, " is not a mystery number.\n")
	}
}

/* Sample I/O:

Enter a number to check if it's a mystery number: 33

The two digits are: 12 21
The number 33 is a mystery number.

*/

File: Java/math/MysteryNumber.java
/* Below program checks if a number is a mystery number. 
Mystery number - A number that can be represented by sum
of two numbers, such that the two numbers are reverse of
each other.*/

import java.util.*;
import java.lang.*;

public class MysteryNumber {

	/* This function takes a number as a input and
	returns	a number with the digits reversed*/
	static int reverse(int number) {

		int rev = 0;
		while(number != 0) {
			rev *= 10;
			rev += number % 10;
			number = number / 10;
		}
		return rev;
	}

	// This function checks if the number is a mystery number
	static int isMystery(int number) {

		// We iterate till the middle of the number
		for(int i = 1; i <= number / 2; i++) {

			// Find reverse of current iteration value
			int rev = reverse(i);
			// Check the sum of number with its reverse
			if(i + rev == number) {
				// If its is equal to the given number we return and print
				System.out.println("\nThe two digits are: " + i + " " + rev);
				return 1;
			}
		}
		// Else return -1
		return -1;
	}

	public static void main(String args[]) {

		// Taking number as input from the user
		Scanner scan = new Scanner(System.in);
		System.out.print("Enter the number: ");
		int number = scan.nextInt();
		scan.close();

		// Calling the mystery function
		int result = isMystery(number);

		// If the result is 1, the number is a mystery number
		if(result == 1) {
			System.out.println("The given number " + number + " is a Mystery number.");
		}

		// Else it is not a mystery number
		else {
			System.out.println("\nThe given number " + number + " is not a Mystery number.");
		}
	}
}

/*
Sample I/O:

1)
Enter the number: 343

The two digits are: 122 221
The given number 343 is a Mystery number.

2)
Enter the number: 3214

The given number 3214 is not a Mystery number.

*/
